rotkehlchen/chain/evm/decoding/balancer/v3/decoder.py

```diff
@@ -0,0 +1,182 @@
+import logging
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any
+
+from eth_abi import decode as decode_abi
+
+from rotkehlchen.assets.utils import CHAIN_TO_WRAPPED_TOKEN
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.constants import ZERO_ADDRESS
+from rotkehlchen.chain.evm.decoding.balancer.constants import BALANCER_LABEL, CPT_BALANCER_V3
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    ActionItem,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.chain.evm.decoding.utils import maybe_reshuffle_events
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_evm_address
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+from rotkehlchen.utils.misc import bytes_to_address
+
+from .constants import (
+    BALANCER_V3_POOL_ABI,
+    LIQUIDITY_ADDED_TOPIC,
+    LIQUIDITY_REMOVED_TOPIC,
+    VAULT_ADDRESS,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Balancerv3CommonDecoder(DecoderInterface):
```

Comment:
Is there nothing common between this and balancer v2? Would it not make sense to abstract some common ones?

Reply:
afaics, they take totally different approaches so there's nothing to abstract in this case.